More filesystem syscalls

mulderr opened the pull request:
So:
* added more fs syscalls
* renamed `stringFromHeap2` to `heapStr` and optimized it a bit (it looks like it's going to be used quite a lot)
* reconfigured browserfs to mount `/` instead of only `/tmp`, apparently this is the default for emscripten so it should make sense

I'm not completely sure about some of those implementations but not sure how to properly test either. For now I'm writing little examples like this and am mostly satisfied if they don't throw:

```c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

int main() {
  int res;
  int fd;

  res = mkdir("/foo", 0700);
  printf("mkdir: %d\n", res);

  fd = creat("/foo/bar", 0600);
  printf("creat: %d\n", fd);
  if (fd > 0)
    close(fd);

  res = rename("/foo/bar", "/foo/baz");
  printf("rename: %d\n", res);

  res = unlink("/foo/baz");
  printf("unlink: %d\n", res);

  res = rmdir("/foo");
  printf("rmdir: %d\n", res);

  return 0;
}
```

mulderr committed:
wasm32: implement SYS_creat

mulderr committed:
wasm32: change browserfs config, add more fs syscalls

ElvishJerricco commented:
This is great, thanks! Two things:

1. Do you know how BrowserFS handles synchronous file IO? More than once I've seen "synchronous" JS APIs implemented by just pegging the CPU at 100% and waiting, basically. But since we're using a WebWorker, we can use `Atomics.wait`, which is much better.
2. I wonder if we should be declaring all these constants by importing the headers from musl. But I suppose this isn't strictly necessary, as I'm sure there's a similar duplication of constants between real kernels and real libc implementations.

mulderr commented:
That probably depends on the filesystem backend. `InMemory` is trivially synchronous because it's really just a [dictionary](https://github.com/jvilk/BrowserFS/blob/master/src/backend/InMemory.ts#L7). Likewise for `LocalStorage` - the underlying API is synchronous. My best guess is that as long as we stick to those two we should be fine for the time being. As to how browserfs deals with actually async backends, unfortunatelly I don't know right now.

Agreed. It would be great not to have to redefine those constants. I'm not sure however, how to refer to musl headers from JS. Is that a thing? Or do you mean like having scripts to generate JS definitions from musl headers?

Edit: actually, since localStorage is not directly available from a WebWorker when we overlay messaging on top, it will probably be async and then we might have a problem. BrowserFS has `WorkerFS` for this but I haven't used it yet.

Edit2: `WorkerFS` seems to be async only, bummer, we may have to write a replacement using `Atomics.wait`

mulderr committed:
wasm32: catch ApiErrors from BrowserFS

ElvishJerricco commented:
Hm. We may want to contribute to BrowserFS to allow synchronous operations from `WorkerFS` (yea, probably using `Atomics.wait`). At the end of the day, I think we want this `libc` to work both on the main thread and on a webworker, with the main thread just throwing errors when it encounters an `Atomics.wait`-based syscall. `WorkerFS` seems like "The Right Way" to do local storage, so it'd be best if it supported the same model out of the box.

I think we can punt on that though. Just getting everything to work via BrowserFS at all is step 1.

P.S. Let me know whenever you want me to merge this. It looks good to me so far.

mulderr commented:
I'd probably hold off on the merge for now if that's ok. I do like all the feedback though!

I want to take some time to try and maybe come up with a more complex example to verify things actually work and see what might be missing - at least from the most common stuff. If you happen to have some interesting/fun ideas for those laying around I'm all ears ;)

ElvishJerricco commented:
@mulderr try building git :P

mulderr commented:
Whoa! I do like your style sir. You're totally trolling me, but it's working. I just went to check what are the direct dependencies of git: curl, zlib, openssl, expat, and libiconv. Nah.

But zlib... gzip? We're probably not that far off in terms of syscalls. Plus it would be interesting to see how it fares against a [pure JS implementation](https://github.com/beatgammit/gzip-js) performance wise. It would be cool to find it's 10 times faster. It might be 10 times slower too.

Anyway, some self contained GNU command line tool sounds like a decent place to start.

mulderr commented:
Quick update. I finally had time to take a stab at it and I changed my mind. It may be a terrible idea.

Automake, during configure, gives:
```
checking host system type... Invalid configuration `wasm32-unknown-unknown-wasm': machine `wasm32-unknown-unknown' not recognized`
```
which I have no idea what to do about.

And trying to make it build with `cmake` would probably be a project on it's own. Pretty much everything includes `configure` generated headers.

But there is a small development. Stream io seems to be broken or rather missing. Trying to `fopen` and `fwrite`:
```
SYS_open 1024 33345 438 undefined undefined undefined
wasm.js:2243 RuntimeError: invalid function
    at __fwritex (wasm-function[67]:69)
    at fwrite (wasm-function[68]:42)
    at main (wasm-function[9]:33)
    at http://localhost:9000/wasm.js:2241:36
    at <anonymous>
```

So I'd like to at least try and fix that.

ElvishJerricco commented:
Re automake: Yea there's this file called config.sub that a lot of projects seem to copy paste from some GNU place. We've had to patch it in a number of projects. [See here](https://github.com/WebGHC/ghc/commit/b9cf10b8d8cfe16b2bd2dd49271ce7c9a6238486).

mulderr commented:
Nice. Fixing config.sub allows it go through that particular check but it errors further down the line. Will retry later.

I need to curb my enthusiasm once again. Stream IO requires malloc and free and those fail for me right now due to missing rt_sigprocmask for starters. 

Btw. is there an easy way to force `-O0` or so with current infrastructure? That would be really helpful for debugging. I'm seeing really weird behavior I suspect is due to optimizations. Like this works:

```c
int main() {
  char *s = malloc(64);
  memset(s, 0x41, 63);
  s[63] = '\0';
  printf("%s\n", s);

  return 0;
}
```

But this doesn't:
```c
int main() {
  char *s = malloc(64);
  memset(s, 0x41, 63);
  s[63] = '\0';
  printf("%s\n", s);
  free(s);

  return 0;
}
```

PS. I should probably mention I'm using all of wasm-cross to compile examples. There's probably an easier way?

ElvishJerricco commented:
Yea nixpkgs does this super annoying things [where it forces -O2 all the time](https://github.com/NixOS/nixpkgs/issues/18995). There's a line in cross.nix that defines hardeningDisable; add "fortify" to that to disable this dumb behavior.

Also, make sure you're setting -mthread-model=single, though the Nix should be doing this for you. I believe that rt_sigprocmask has something to do with multithreading, which we don't have. It's possible we need to actually disable the building of pthread functions in musl. Right now I think we are building them (and just having them fail due to unimplemented syscalls), which confuses configure scripts that check if pthread functions are declared.

EDIT: Also note that we can't currently release memory back to the browser. `free` should be able to work though; it just shouldnt succeed in reducing the size of the memory.

mulderr commented:
[This](https://github.com/WebGHC/musl/blob/syscalls-js/arch/wasm32/atomic_arch.h#L9) is where `free` ends up. I don't know what it is but it's not `abort` on any of the other architectures. It's the reason for 100% CPU - we're still SIGKILL proof and `abort` is send some signals plus `for(;;)`.

ElvishJerricco commented:
Yea this is confirming my suspicion that our musl is pretending to support pthreads threading. I think we need to figure out how to get it not to build any threading and atomic stuff.

ElvishJerricco commented:
Also, I believe `exit` and `abort` just send some syscalls, then loop indefinitely, I guess waiting for the OS to asynchronously kill it. Those syscalls ought to be triggering out `NYI` exceptions though, so I'm not sure why it would be going into the loop.

mulderr commented:
You're right. It should throw NYI! I commented that out in `tkill` at one point for some reason. So yeah. `abort` raises SIGKILL and that kills it which is kinda backwards because the exception is thrown upon trying to send the signal. It should be sane to keep it that way since the whole thing is trapped in a single process/thread so assuming the arguments are correct it can only send SIGKILL to itself. I think.

mulderr committed:
wasm32: define a_cas to make malloc work (fixes #5)

mulderr committed:
wasm32: Add errno.js, create some POSIX required directories

mulderr committed:
wasm32: readv and writev in terms of read and write

mulderr committed:
Merge branch 'syscalls-js' of https://github.com/WebGHC/musl into syscalls-js

mulderr committed:
wasm32: use errno.js in stime

mulderr commented:
Added a bunch of things, most importantly:
* `a_cas` definition that makes `malloc` work
* `errno.js` with all possible error values - hope, that's ok, if not just shout and I'll remove it
* added an initialization function to create some standard POSIX directories
* implemented `readv` and `writev` in terms of `read` and `write` which makes simple stream IO work
* moved some stuff to `util.js` because I ended up needing it in both `wasm.js` and `fs.js`

Btw. since my minimal goal (stream IO) is sort of done, if you're ok with these additional changes, I'd rather close this now and tackle those other ideas in new PRs - this is slowly getting messy. If not, I'm happy to fix any issues.

ElvishJerricco commented on arch/wasm32/atomic_arch.h:
```diff
@@ -6,6 +6,13 @@
 #define a_barrier() syscall(__NR_membarrier)
 
-#define a_cas(p, t, s) (abort(), s)
+#define a_cas a_cas
```
What is the `#define a_cas a_cas` for? AFAIK, having this is identical to not having it.

mulderr replied:
That's needed if you later want to test with `#ifdef`. Just defining a function with that same name will not work - this does not print anything:
```c
#include <stdio.h>

int const7()
{
  return 7;
}

int
main()
{
#ifdef const7
  printf("%d\n", const7());
#endif

  return 0;
}
```
But it will work if you add `#define const7 const7`. I don't want to dig into C standards right now but I guess function namespace is separate from preprocessor namespace would be the way to say it more formally?

ElvishJerricco replied:
Ah. Alright, looks good to me.

mulderr replied:
While we're at it we may want to add a comment saying it's not really a compare-and-swap - it's not atomic. But then it's ok as long as we're still single threaded. AFAIK WASM MVC does not include atomic operations of this kind so we may want to revisit at a later time. They do [say](https://github.com/WebAssembly/design/blob/master/Portability.md) though that support for CAS is required in the execution environment - they call it compare-and-exchange (apparently it goes by many names). So it's very likely it will be available at some point.

ElvishJerricco replied:
~~Yea if you don't mind just leaving a comment describing the situation. We should note that there is a threading proposal which will allow us to implement this properly, but without threading it isn't necessary.~~ I got it :P

ElvishJerricco commented:
Great call on the `errno.js` thing. Is it possible to generate that with a sed script or something? If so, it'd be good to have that in a comment.
